shared/chat/channel-mention-hud/index.js

```diff
@@ -0,0 +1,146 @@
+const Hud = ({style, data, rowRenderer, selectedIndex}: Props<*>) =>
+  data.length ? (
+    <Box style={{...hudStyle, ...style}}>
+      <List items={data} renderItem={rowRenderer} selectedIndex={selectedIndex} fixedHeight={40} />
+    </Box>
+  ) : null
+
+const hudStyle = {
+  ...globalStyles.flexBoxRow,
+  backgroundColor: globalColors.white,
+}
+
+type MentionHudProps = {
+  channels: Array<{channelName: string}>,
+  onPickChannel: (user: string) => void,
+  onSelectChannel: (user: string) => void,
+  pickSelectedChannelCounter: number,
+  selectUpCounter: number,
+  selectDownCounter: number,
+  filter: string,
+  style?: ?Object,
+}
+
+// TODO figure typing out
+const mapStateToProps: MapStateToProps<*, *, *> = (state: TypedState, {channels, selectedIndex, filter}) => ({
+  data: channels
+    ? Object.keys(channels)
+        .filter(c => c.toLowerCase().indexOf(filter) >= 0)
+        .sort()
+        .map((c, i) => ({channelName: c, selected: i === selectedIndex}))
+    : {},
+})
+
+// TODO share this connector with user-mention-hud?
+// $FlowIssue is confused
+const MentionHud: Class<React.Component<MentionHudProps, void>> = compose(
+  withState('selectedIndex', 'setSelectedIndex', 0),
+  connect(mapStateToProps),
+  lifecycle({
+    componentWillReceiveProps: function(nextProps) {
+      if (nextProps.data.length === 0) {
+        nextProps.setSelectedIndex(0)
+      }
+      if (nextProps.data.length && nextProps.data.length !== this.props.data.length) {
+        nextProps.setSelectedIndex(n => Math.min(n, nextProps.data.length - 1))
+      }
+
+      if (nextProps.selectUpCounter !== this.props.selectUpCounter) {
+        nextProps.setSelectedIndex(n => {
+          const next = n - 1
+          if (next < 0) {
+            return Math.max(nextProps.data.length - 1, 0)
+          }
+          return next
+        })
+      } else if (nextProps.selectDownCounter !== this.props.selectDownCounter) {
+        nextProps.setSelectedIndex(n => {
+          const next = n + 1
+          if (next >= nextProps.data.length) {
+            return 0
+          }
+          return next
+        })
+      }
+
+      if (nextProps.pickSelectedChannelCounter !== this.props.pickSelectedChannelCounter) {
+        if (nextProps.selectedIndex < nextProps.data.length) {
+          nextProps.onPickChannel(nextProps.data[nextProps.selectedIndex].channelName)
+        } else {
+          // Just exit
+          nextProps.onPickChannel(nextProps.filter, {notChannel: true})
+        }
+      }
+
+      if (nextProps.selectedIndex !== this.props.selectedIndex) {
+        if (nextProps.selectedIndex < nextProps.data.length) {
+          nextProps.onSelectChannel(nextProps.data[nextProps.selectedIndex].channelName)
+        }
+      }
+    },
+  }),
+  withPropsOnChange(['onPickChannel'], ownerProps => ({
+    rowRenderer: (index, props) => {
+      return (
+        <MentionRowRenderer
+          key={props.key}
```

Comment:
it looks like `props.key` is undefined here, I get the react array key error when I enter `#` in a team channel

Reply:
@buoyad Thanks! Fixed that too.